### DISCUSSION TITLE
how to correctly modify surface normal in fragment shader

### DISCUSSION THREAD

**Discussion by JackKaiXing** [google/filament#4778] [Q&A] [answered]
Hello,

I want to do some normal bump in shader as time changes.

```glsl
fragment {
     void material(inout MaterialInputs material) {
          preparematerial(material);
          // ....
          material.normal = MyTimeChangedNormal;
          // ....
     }
}
```

Does it correct? Thanks

**Comment by romainguy** [comment, marked as answer]
You need to modify the normal before calling prepareMaterial().

**Reply by JackKaiXing** [reply]
@romainguy 

Yes, it works. 

But another question is that "if the material is `lit` and I pre-calculate `tangents(using packTangentFrame)` for vertex attribute, when the normal is changed in fragment shader, how about the tangents and bitangents?".

Should I change the `bitangent and tangent`?

**Reply by romainguy** [reply]
It depends on what you are trying to do but the normal you supply will use the frame defined by the geometric normal + tangent and bitangent to be properly transformed.

**Reply by romainguy** [reply]
I.e. don’t touch anything :)

**Reply by JackKaiXing** [reply]
@romainguy 

Thanks for your quick reply.

I has a triangle and pre-calculated its normal/bitangent/tangent, and I also has a normal bump texture. for my lit shader, I use the sampled value of bump texture to modify the normal. 

But I need lighting calculation is correct with the new normal. According to you last answer, if I do not change the tangent frame, will the new normal be transformed correctly?

**Reply by JackKaiXing** [reply]
@romainguy 

If vertexes of neighboring triangles are in the same plane(i.e. same normal) with however different tangent/bitangent(maybe opposite directions), the the line between two triangles is obvious after normal bump, which I guess is because of the different tangent frame.
